### ISSUE TITLE
Add PodSecurityPolicy admission information to audit logs

### ISSUE THREAD

**Issue by crassirostris** [kubernetes/kubernetes#56209]
Follow-up of https://github.com/kubernetes/kubernetes/pull/55486#issuecomment-343873447

Quoting @sttts:

> The PSP admission plugin matches a pod with a PodSecurityPolicy. If no policy matches, the pod is rejected. It would be helpful to know which policy made the pod pass the admission step.

/cc @tallclair @sttts @kubernetes/sig-auth-feature-requests @kubernetes/sig-instrumentation-feature-requests

**Comment by tallclair** [comment]
I don't think this should be baked into the audit log API, since the PSP controller is an optional plugin. Instead, I think we should add some extension mechanisms to the audit log, and extend the admission API to expose that. E.g. `AdmissionAnnotations` or just use the existing annotations.

I could see a similar use for authentication & authorization. E.g.:
- Request was authenticated using the webhook plugin (or maybe even let the webhook add arbitrary annotations, such as an internal user id)
- Request was authorized using RBAC, ClusterRole=foo RoleBinding=bar
- PodSecurityPolicy admitted using PSP baz
- etc.

**Comment by sttts** [comment]
> I think we should add some extension mechanisms to the audit log, and extend the admission API to expose that. E.g. AdmissionAnnotations or just use the existing annotations.

Totally agree. But we should not use normal annotations for that. This is sensible information possibly and also no something we want to carry around on objects. Moreover, we want this information not only for mutating API calls, but also for read-only requests.

**Comment by tallclair** [comment]
@sttts to clarify, I was talking about the annotations on the `audit.Event` object, except I forgot that we removed the `ObjectMeta` from that in the beta API, so there no longer is an `Annotations` field.

I guess the question is, do we want to have `AdmissionAnnotations`, `AuthenticationAnnotations`, and `AuthorizationAnnotations` fields on the `audit.Event`, or should those all share a single `Annotations` field? (I'm leaning towards the latter).

**Comment by sttts** [comment]
I am fine with either.

Just wondering whether annotation is the right term. Usually, annotations are for tooling, nothing to rely one. Maybe annother name for the field?

**Comment by CaoShuFeng** [comment]
Adding AdmissionAnnotations and AuthenticationAnnotations will be duplication of this:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/audit/request.go#L176

What's your suggestions?

**Comment by sttts** [comment]
> Adding AdmissionAnnotations and AuthenticationAnnotations will be duplication of this:

Cannot follow.The status is a special response type we handle in a special way. Here we talk about meta information of the admission pipeline that we want to have in audit events. It's not about "normal annotation" of objects.

**Comment by CaoShuFeng** [comment]
> It would be helpful to know which policy made the pod pass the admission step.

Oh, I get it.
I thought I was "which psp rejected the request."
But what we want is "which psp accepted the request."

**Comment by CaoShuFeng** [comment]
Hi:
Api changes to advanced audit:   https://github.com/kubernetes/kubernetes/pull/58806
add rbac info to advanced audit: https://github.com/kubernetes/kubernetes/pull/58807
add psp info to advanced audit : https://github.com/kubernetes/kubernetes/pull/58143
support annotations for admission webhook: https://github.com/kubernetes/kubernetes/pull/58679 (WIP)

**Comment by tallclair** [comment]
Thanks for breaking this up, it's much easier to review now.

**Comment by tallclair** [comment]
Sorry, meant to close https://github.com/kubernetes/kubernetes/pull/58143.
